File: src/main/java/org/jdc/template/ui/IndividualEditFragment.java
package org.jdc.template.ui;

import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.google.android.gms.analytics.HitBuilders;

import org.apache.commons.lang3.StringUtils;
import org.jdc.template.Analytics;
import org.jdc.template.App;
import org.jdc.template.R;
import org.jdc.template.dagger.Injector;
import org.jdc.template.domain.main.individual.Individual;
import org.jdc.template.domain.main.individual.IndividualManager;
import org.jdc.template.event.IndividualSavedEvent;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;
import de.greenrobot.event.EventBus;
import pocketknife.BindArgument;
import pocketknife.PocketKnife;

public class IndividualEditFragment extends Fragment {
    public static final String TAG = App.createTag(IndividualEditFragment.class);

    private static final String ARG_ID = "ID";

    @Bind(R.id.first_name_layout)
    TextInputLayout firstNameLayout;

    @Bind(R.id.first_name)
    EditText firstNameEditText;

    @Bind(R.id.last_name)
    EditText lastNameEditText;

    @Bind(R.id.phone)
    EditText phoneEditText;

    @Bind(R.id.email)
    EditText emailEditText;

    @Inject
    IndividualManager individualManager;

    @Inject
    EventBus bus;

    @Inject
    Analytics analytics;

    @BindArgument(ARG_ID)
    long individualId;

    public static IndividualEditFragment newInstance(long individualId) {
        IndividualEditFragment fragment = new IndividualEditFragment();
        Bundle args = new Bundle();
        args.putLong(ARG_ID, individualId);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Injector.get().inject(this);
        PocketKnife.bindArguments(this);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(@Nonnull LayoutInflater inflater, ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View view = inflater.inflate(R.layout.fragment_individual_edit, container, false);
        ButterKnife.bind(this, view);

        showIndividual();

        return view;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.individual_edit_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_item_save:
                saveIndividual();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void showIndividual() {
        if (individualId <= 0) {
            return;
        }

        Individual individual = individualManager.findByRowId(individualId);
        if (individual != null) {
            analytics.send(new HitBuilders.EventBuilder()
                    .setCategory(Analytics.CATEGORY_INDIVIDUAL)
                    .setAction(Analytics.ACTION_EDIT)
                    .build());

            firstNameEditText.setText(individual.getFirstName());
            lastNameEditText.setText(individual.getLastName());
            emailEditText.setText(individual.getEmail());
            phoneEditText.setText(individual.getPhone());
        }
    }

    private void saveIndividual() {
        Individual individual = individualId > 0 ? individualManager.findByRowId(individualId) : new Individual();
        if (individual != null) {
            if (StringUtils.isBlank(firstNameEditText.getText())) {
                firstNameLayout.setError(getString(R.string.required));
                return;
            }

            individual.setFirstName(firstNameEditText.getText().toString());
            individual.setLastName(lastNameEditText.getText().toString());
            individual.setPhone(phoneEditText.getText().toString());
            individual.setEmail(emailEditText.getText().toString());

            individualManager.save(individual);

            bus.post(new IndividualSavedEvent(individual.getId()));
        }
    }
}

File: src/main/kotlin/org/jdc/template/ui/IndividualEditFragment.kt
package org.jdc.template.ui

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.os.Bundle
import android.support.v4.app.Fragment
import android.text.format.DateUtils
import android.view.*
import com.google.android.gms.analytics.HitBuilders
import kotlinx.android.synthetic.main.fragment_individual_edit.*
import org.apache.commons.lang3.StringUtils
import org.dbtools.android.domain.DBToolsDateFormatter
import org.jdc.template.Analytics
import org.jdc.template.App
import org.jdc.template.R
import org.jdc.template.R.layout.fragment_individual_edit
import org.jdc.template.dagger.Injector
import org.jdc.template.domain.main.individual.Individual
import org.jdc.template.domain.main.individual.IndividualManager
import org.jdc.template.event.IndividualSavedEvent
import org.jdc.template.event.RxBus
import org.threeten.bp.LocalDate
import org.threeten.bp.LocalTime
import org.threeten.bp.ZoneId
import pocketknife.BindArgument
import pocketknife.PocketKnife
import javax.inject.Inject

class IndividualEditFragment : Fragment() {

    @Inject
    lateinit var individualManager: IndividualManager
    @Inject
    lateinit var bus: RxBus
    @Inject
    lateinit var analytics: Analytics

    @BindArgument(ARG_ID)
    var individualId: Long = 0

    private var individual = Individual()

    private var birthDatePickerDialog: DatePickerDialog? = null
    private var alarmTimePickerDialog: TimePickerDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Injector.get().inject(this)
        PocketKnife.bindArguments(this)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(fragment_individual_edit, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        alarmTimeEditText.setOnClickListener() {
            onAlarmClick()
        }

        birthDateEditText.setOnClickListener() {
            onBirthdayClick()
        }

        showIndividual()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.individual_edit_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.menu_item_save -> {
                saveIndividual()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    fun onBirthdayClick() {
        if (birthDatePickerDialog == null) {

            val date = if (individual.birthDate != null) individual.birthDate else LocalDate.now()
            if (date != null) {
                birthDatePickerDialog = DatePickerDialog(activity, DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
                    individual.birthDate = LocalDate.of(year, monthOfYear + 1, dayOfMonth) // + 1 because cord Java Date is 0 based
                    showBirthDate()
                }, date.getYear(), date.getMonthValue() - 1, date.getDayOfMonth()) // - 1 because cord Java Date is 0 based
            }
        }

        birthDatePickerDialog!!.show()
    }

    fun onAlarmClick() {
        if (alarmTimePickerDialog == null) {
            val time = if (individual.alarmTime != null) individual.alarmTime else LocalTime.now()
            alarmTimePickerDialog = TimePickerDialog(activity, TimePickerDialog.OnTimeSetListener { view, hourOfDay, minute ->
                individual.alarmTime = LocalTime.of(hourOfDay, minute)
                showAlarmTime()
            }, time.hour, time.minute, false)
        }

        alarmTimePickerDialog!!.show()
    }

    private fun showIndividual() {
        if (individualId <= 0) {
            return
        }

        val foundIndividual = individualManager.findByRowId(individualId)
        if (foundIndividual != null) {
            individual = foundIndividual
            analytics.send(HitBuilders.EventBuilder().setCategory(Analytics.CATEGORY_INDIVIDUAL).setAction(Analytics.ACTION_EDIT).build())

            firstNameEditText.setText(individual.firstName)
            lastNameEditText.setText(individual.lastName)
            emailEditText.setText(individual.email)
            phoneEditText.setText(individual.phone)

            showBirthDate()
            showAlarmTime()
        }
    }

    private fun showBirthDate() {
        if (individual.birthDate == null) {
            return
        }

        val date = individual.birthDate
        if (date != null) {
            val millis = DBToolsDateFormatter.localDateTimeToLong(date.atStartOfDay(ZoneId.systemDefault()).toLocalDateTime())
            birthDateEditText.setText(DateUtils.formatDateTime(activity, millis!!, DateUtils.FORMAT_SHOW_DATE or DateUtils.FORMAT_SHOW_YEAR))
        }
    }

    private fun showAlarmTime() {
        if (individual.alarmTime == null) {
            return
        }

        val time = individual.alarmTime
        val millis = DBToolsDateFormatter.localDateTimeToLong(time.atDate(LocalDate.now()))
        alarmTimeEditText.setText(DateUtils.formatDateTime(activity, millis!!, DateUtils.FORMAT_SHOW_TIME))
    }

    private fun saveIndividual() {
        if (StringUtils.isBlank(firstNameEditText.text)) {
            firstNameLayout.error = getString(R.string.required)
            return
        }

        if (individual.alarmTime == null) {
            alarmTimeLayout.error = getString(R.string.required)
            return
        }

        individual.firstName = firstNameEditText.text.toString()
        individual.lastName = lastNameEditText.text.toString()
        individual.phone = phoneEditText.text.toString()
        individual.email = emailEditText.text.toString()

        individualManager.save(individual)

        bus.post(IndividualSavedEvent(individual.id))
        bus.post(IndividualSavedEvent(individual.id))
    }

    companion object {
        val TAG = App.createTag(IndividualEditFragment::class.java)

        const val ARG_ID = "ID"

        fun newInstance(individualId: Long): IndividualEditFragment {
            val fragment = IndividualEditFragment()
            val args = Bundle()
            args.putLong(ARG_ID, individualId)
            fragment.arguments = args
            return fragment
        }
    }
}
